src/vidpf.rs

```diff
@@ -0,0 +1,667 @@
+        rng.fill_bytes(&mut out_seed.0 .0);
+        let value = R::from_seed_stream(&mut rng);
+
+        (out_seed, value)
+    }
+
+    fn hash_one(alpha: &[u8], level: usize, seed: &Seed) -> Proof {
+        let dst = b"vidpf cs proof";
+        let mut binder = Vec::new();
+        binder.extend_from_slice(alpha);
+        binder.extend(level.to_le_bytes());
+        let mut rng = XofTurboShake128::seed_stream(&seed.0, dst, &binder);
+
+        let mut proof = Proof::default();
+        rng.fill_bytes(&mut proof.0);
+
+        proof
+    }
+
+    fn hash_two(proof: &Proof) -> Proof {
+        let dst = b"vidpf proof adjustment";
+        let binder = &proof.0;
+        let seed = Seed::default();
+        let mut rng = XofTurboShake128::seed_stream(&seed.0, dst, binder);
+
+        let mut out_proof = Proof::default();
+        rng.fill_bytes(&mut out_proof.0);
+
+        out_proof
+    }
+}
+
+/// VIDPF-related errors.
+#[derive(Debug, thiserror::Error)]
+#[non_exhaustive]
+pub enum VidpfError {
+    /// Invalid parameter, indicates an invalid input.
+    #[error("invalid parameter: {0}")]
+    InvalidParameter(String),
+}
+
+/// `ServerID` used to identify the two aggregation servers.
+#[derive(Debug, Clone, Copy)]
+pub enum ServerID {
+    /// S0 is the first server.
+    S0 = 0,
+    /// S1 is the second server.
+    S1 = 1,
+}
+
+#[derive(Debug)]
+/// Share
+pub struct Share<R>
+where
+    R: Range,
+    for<'a> &'a R: RangeOps<R>,
+{
+    /// y is the sharing of the output.
+    pub y: R,
+    /// pi is a proof used to verify the share.
+    pub pi: Proof,
+}
+
+/// Represents the minimum number of arithmetic operations of the
+/// codomain of a Distributed Point Function (DPF).
+pub trait RangeOps<R>: Sized + Add<Output = R> + Sub<Output = R> {}
+
+/// Represents the codomain of a Distributed Point Function (DPF).
+pub trait Range: PartialEq + ConditionallyNegatable + ConditionallySelectable
+where
+    for<'a> &'a Self: RangeOps<Self>,
+{
+    /// Creates a new value initialized to zero.
+    fn new() -> Self;
+    /// Returns true if the value is zero.
+    fn is_zero(self) -> bool;
+    /// Fills the value from a pseudorandom generator.
+    fn from_seed_stream<R: RngCore>(rng: R) -> Self;
+}
+
+#[derive(Debug)]
+/// `CorrectionWord` adjusts values of shares during VIDPF evaluation.
+pub struct CorrectionWord<R>
+where
+    R: Range,
+    for<'a> &'a R: RangeOps<R>,
+{
+    s_cw: Seed,
+    t_cw_l: ControlBit,
+    t_cw_r: ControlBit,
+    w_cw: R,
+}
+
+#[derive(Debug, PartialEq, Eq, Default)]
+/// Proof
+pub struct Proof([u8; 32]);
+
+impl BitXor for &Proof {
+    type Output = Proof;
+
+    fn bitxor(self, rhs: Self) -> Self::Output {
+        let mut out = Proof::default();
+        zip(&mut out.0, zip(self.0, rhs.0)).for_each(|(c, (a, b))| *c = a ^ b);
+
+        out
+    }
+}
+
+#[derive(Debug)]
+/// `PublicKey` is the common public key of aggregation servers.
+pub struct PublicKey<R>
```

Comment:
This is called the "public share" in IDPF parlance.
```suggestion
/// `PublicShare` is the common public key of aggregation servers.
pub struct PublicShare<R>
```